dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/dataelement/DataElementMergeService.java

```diff
@@ -0,0 +1,226 @@
+  private final TrackerDataElementMergeHandler trackerMergeHandler;
+  private final CommonMergeHandler commonMergeHandler;
+  private final MergeValidator validator;
+  private final DataElementMergeValidator dataElementMergeValidator;
+  private final EntityManager entityManager;
+  private ImmutableList<DataElementMergeHandler> metadataMergeHandlers;
+  private ImmutableList<MergeHandler> commonMergeHandlers;
+  private ImmutableList<DataElementDataMergeHandler> dataMergeHandlers;
+  private ImmutableList<DataElementAuditMergeHandler> auditMergeHandlers;
+
+  @Override
+  public MergeRequest validate(@Nonnull MergeParams params, @Nonnull MergeReport mergeReport) {
+    log.info("Validating DataElement merge request");
+    if (params.getDataMergeStrategy() == null) {
+      mergeReport.addErrorMessage(new ErrorMessage(ErrorCode.E1556));
+      return null;
+    }
+
+    // sources
+    Set<UID> sources = new HashSet<>();
+    validator.verifySources(params.getSources(), sources, mergeReport, DataElement.class);
+
+    // target
+    validator.checkIsTargetInSources(sources, params.getTarget(), mergeReport, DataElement.class);
+    MergeRequest request = validator.verifyTarget(mergeReport, sources, params, DataElement.class);
+
+    if (mergeReport.hasErrorMessages()) return request;
+
+    // get DEs for further type-specific validation
+    DataElement deTarget = dataElementService.getDataElement(request.getTarget().getValue());
+    List<DataElement> deSources =
+        dataElementService.getDataElementsByUid(
+            request.getSources().stream().map(UID::getValue).toList());
+
+    MergeReport report =
+        dataElementMergeValidator.validateValueType(deTarget, deSources, mergeReport);
+
+    if (report.hasErrorMessages()) return request;
+
+    dataElementMergeValidator.validateDomainType(deTarget, deSources, mergeReport);
+    return request;
+  }
+
+  @Override
+  @Transactional
+  public MergeReport merge(@Nonnull MergeRequest request, @Nonnull MergeReport mergeReport) {
+    log.info("Performing DataElement merge");
+
+    List<DataElement> sources =
+        dataElementService.getDataElementsByUid(UID.toValueList(request.getSources()));
+    DataElement target = dataElementService.getDataElement(request.getTarget().getValue());
+
+    // merge metadata
+    log.info("Handling DataElement reference associations and merges");
+    dataMergeHandlers.forEach(h -> h.merge(sources, target, request));
+    metadataMergeHandlers.forEach(h -> h.merge(sources, target));
+    commonMergeHandlers.forEach(h -> h.merge(sources, target));
+    auditMergeHandlers.forEach(h -> h.merge(sources, request));
+
+    // a flush is required here to bring Hibernate into a consistent state, due to some of the
+    // above merge operations involving required native queries (queries involving JSONB). This
+    // eliminates possible stale state issues.
+    entityManager.flush();
```

Comment:
Maybe you can utilize `FlushMode` for each query in order to make sure the query will not cause stale data.

Reply:
hmm, possibly. So it is currently set up to flush (once) at the end of all merges, which seems to work nicely. All DB constraints are adhered to and the native queried data behaves as expected  (in tests anyway).
Do you think flushing for some/all queries would be better somehow?

Reply:
- So the default FlushMode is AUTO, meaning we don't know when hibernate will flush the session.
- Therefore, we can't know if calling flush explicitly in this point is good or bad, depends on when is the last time hibernate flush the session.
- Another factor to consider flushing explicitly is when you have a very big list of objects in the persistent context. In that case, we would want implement batch process.
- As this is a new feature, I think we can wait for the feedbacks and see if there is any issue with this one.